File: backend/test/python/demo_python_vulnerable_project/safe/safe_examples.py
# safe/safe_examples.py

def safe_find_user(request, cursor):
    user_id = request.GET["id"]

    query = "SELECT * FROM users WHERE id = %s"

    cursor.execute(query, (user_id,))


def safe_update_order(request, cursor):
    order_id = request.POST["order_id"]
    status = request.POST["status"]

    query = "UPDATE orders SET status = %s WHERE id = %s"

    cursor.execute(query, (status, order_id))


def safe_static_command():
    command = "echo backup started"

    return command


def safe_mongo_query(request, users_collection):
    username = request.GET["username"]

    query = {
        "username": username
    }

    result = users_collection.find(query)

    return result

File: backend/test/php/demo_php_vulnerable_project/safe/safe_examples.php
<?php
// safe/safe_examples.php

function safe_find_user($connection) {
    $userId = $_GET["id"];

    $query = "SELECT * FROM users WHERE id = ?";

    $stmt = $connection->prepare($query);
    $stmt->bind_param("s", $userId);
    $stmt->execute();
}


function safe_update_order($connection) {
    $orderId = $_POST["order_id"];
    $status = $_POST["status"];

    $query = "UPDATE orders SET status = ? WHERE id = ?";

    $stmt = $connection->prepare($query);
    $stmt->bind_param("ss", $status, $orderId);
    $stmt->execute();
}


function safe_static_command() {
    $command = "echo backup started";

    return $command;
}


function safe_single_quoted_literal($connection) {
    $userId = $_GET["id"];

    // In PHP single quotes do not interpolate variables.
    $query = 'SELECT * FROM users WHERE id = $userId';

    mysqli_query($connection, $query);
}
?>